erigon-lib/downloader/downloader.go

```diff
@@ -575,39 +614,232 @@ func (d *Downloader) mainLoop(silent bool) error {
+				case <-time.After(10 * time.Second):
+					continue
 				}
-				t.DownloadAll()
-				d.wg.Add(1)
-				go func(t *torrent.Torrent) {
-					defer d.wg.Done()
-					defer sem.Release(1)
-					select {
-					case <-d.ctx.Done():
-						return
-					case <-t.Complete.On():
+			}
+
+			d.lock.RLock()
+			downloadingLen := len(d.downloading)
+			d.stats.Downloading = int32(downloadingLen)
+			d.lock.RUnlock()
+
+			available := availableTorrents(d.ctx, pending, d.cfg.DownloadSlots-downloadingLen)
+
+			availableLen := len(available)
+			addedWeb := 0
+			replacedWeb := 0
+
+			for _, webDownload := range d.webDownloadInfo {
+				d.lock.RLock()
+				_, downloading := d.downloading[webDownload.torrent.Name()]
+				d.lock.RUnlock()
+
+				if downloading {
+					continue
+				}
+
+				addDownload := true
+
+				for _, t := range available {
+					if t.Name() == webDownload.torrent.Name() {
+						addDownload = false
+						break
+					}
+				}
+
+				if addDownload {
+					if len(available) < d.cfg.DownloadSlots-downloadingLen {
+						addedWeb++
+						available = append(available, webDownload.torrent)
 					}
-				}(t)
+				} else {
+					wi, _ := snaptype.ParseFileName(d.SnapDir(), webDownload.torrent.Name())
+
+					for i, t := range available {
+						ai, _ := snaptype.ParseFileName(d.SnapDir(), t.Name())
+
+						if ai.CompareTo(wi) > 0 {
+							replacedWeb++
+							available[i] = webDownload.torrent
+							break
+						}
+					}
+				}
 			}
 
-			select {
-			case <-d.ctx.Done():
-				return
-			case <-time.After(10 * time.Second):
+			if len(pending) > 0 {
+				d.logger.Debug("available", "torrents", len(torrents), "pending", len(pending), "available", availableLen, "web-added", addedWeb, "web-replaced", replacedWeb)
+			}
+
+			for _, t := range available {
+				if err := sem.Acquire(d.ctx, 1); err != nil {
+					d.logger.Warn("Failed to acquire download semaphore", "err", err)
+					return
+				}
+
+				switch {
+				case len(t.PeerConns()) > 0:
+					d.logger.Debug("[snapshots] Downloading from torrent", "file", t.Name(), "peers", len(t.PeerConns()))
+					d.torrentDownload(t, sem)
+				case len(t.WebseedPeerConns()) > 0:
+					if d.webDownloadClient != nil {
+						var peerUrls []*url.URL
+
+						for _, peer := range t.WebseedPeerConns() {
+							if peerUrl, err := webPeerUrl(peer); err == nil {
+								peerUrls = append(peerUrls, peerUrl)
+							}
+						}
+
+						d.logger.Debug("[snapshots] Downloading from web", "file", t.Name(), "peers", len(t.WebseedPeerConns()))
+						session, err := d.webDownload(peerUrls, t, nil, webDownloadComplete, sem)
+
+						if err != nil {
+							d.logger.Warn("Can't complete web download", "file", t.Info().Name, "err", err)
+
+							if session == nil {
+								d.torrentDownload(t, sem)
+							}
+
+							continue
+						}
+
+					} else {
+						d.torrentDownload(t, sem)
+					}
+				default:
+					if d.webDownloadClient != nil {
+						webDownload, ok := d.webDownloadInfo[t.Name()]
```

Comment:
forgotten (?) mutex

Reply:
I'll look at tiding these up in the morning.  I think at the moment this map is only used in the main loop.  But its worth protecting it now it moved to be managed by the object not the function.